cups/file.c

```diff
@@ -335,6 +335,135 @@ _cupsFileCheckFilter(
 #endif /* !_WIN32 */
 
 
+/*
+ * '_cupsFileGetConfAndComments()' - Get line and comments from a configuration file.
+ */
+
+char *					/* O  - Line read or @code NULL@ on end of file or error */
+_cupsFileGetConfAndComments(cups_file_t *fp,	/* I  - CUPS file */
+                char        *buf,	/* O  - String buffer */
+		size_t      buflen,	/* I  - Size of string buffer */
+                char        **value,	/* O  - Pointer to value */
+		int         *linenum)	/* IO - Current line number */
+{
+  char	*ptr;				/* Pointer into line */
+
+
+ /*
+  * Range check input...
+  */
+
+  DEBUG_printf(("2cupsFileGetConfAndComments(fp=%p, buf=%p, buflen=" CUPS_LLFMT
+                ", value=%p, linenum=%p)", (void *)fp, (void *)buf, CUPS_LLCAST buflen, (void *)value, (void *)linenum));
+
+  if (!fp || (fp->mode != 'r' && fp->mode != 's') ||
+      !buf || buflen < 2 || !value)
+  {
+    if (value)
+      *value = NULL;
+
+    return (NULL);
+  }
+
+ /*
+  * Read the next line...
+  */
+
+  *value = NULL;
+
+  while (cupsFileGets(fp, buf, buflen))
+  {
+    (*linenum) ++;
+
+    /*
+     * Remove the inline comment...
+     */
+     if ((ptr = strchr(buf, '#')) != NULL)
+     {
+        while(ptr > buf)
+        {
+         // Null-terminate the string after the last non-whitespace, unless the '#' character is escaped by a backslash ('\')
+         if(!_cups_isspace(ptr[-1]) && (ptr == buf || ptr[-1] != '\\'))
+         {
+           *ptr = '\0';
+           break;
+         }
+         ptr--;
+        }
+     }
+
+     /*
+      * Strip leading whitespace...
+      */
+
+      for (ptr = buf; _cups_isspace(*ptr); ptr ++);
+
+      if (ptr > buf)
+        _cups_strcpy(buf, ptr);
+
+       /*
+        * Return the comment if any...
+        */
+        if(buf[0])
+        {
+          if (buf[0] == '#')
```

Comment:
I've thought a bit about dealing with the empty lines in the yesterday evening. We can return ```buf``` here together with ```buf[0] == '#'```, so the condition would look like:

```
/*
 * If we got empty line or comment, return it...
 */
if (!buf[0] || buf[0] == '#')
  return buf;
```

and we can remove the ```if(buf[0])``` before that.

We just need to add a new ```if else if``` in cupsAdminSetServerSettings():

``` 
while (_cupsFileGetConfandComments(cupsd, line, sizeof(line), &value, &linenum))
{
 /*
  *  Preserve empty lines
  */
  if (!line[0])
    cupsFilePuts(temp, "\n");
  else if ((!_cups_strcasecmp(line, "Port") || !_cups_strcasecmp(line, "Listen")) &&
```

This should do the trick - WDYT? I guess you might show something similar in the past, but I didn't get it in the past, sorry for that :D .